File: src/main/kotlin/ghistabs/StabsPlugin.java
package ghistabs;

import docking.ActionContext;
import docking.action.DockingAction;
import docking.action.MenuData;
import ghidra.app.plugin.PluginCategoryNames;
import ghidra.app.plugin.ProgramPlugin;
import ghidra.app.plugin.core.analysis.AutoAnalysisManager;
import ghidra.framework.main.AnalyzerPluginPackage;
import ghidra.framework.plugintool.PluginInfo;
import ghidra.framework.plugintool.PluginTool;
import ghidra.framework.plugintool.util.PluginStatus;
import ghidra.program.model.listing.Program;
import ghidra.util.Msg;

/**
 * Adds a 'Tools > Stabs > Re-import' action that clears the persistent done-flag
 * and re-runs the StabsAnalyzer on the current program.
 */
@PluginInfo(
    status = PluginStatus.STABLE,
    packageName = AnalyzerPluginPackage.NAME,
    category = PluginCategoryNames.ANALYSIS,
    shortDescription = "Re-run the STABS importer on the current program.",
    description = "Adds a 'Tools > Stabs > Re-import' action that clears the persistent done-flag and re-runs the StabsAnalyzer."
)
public class StabsPlugin extends ProgramPlugin {

    public StabsPlugin(PluginTool tool) {
        super(tool);
        DockingAction reimport = new DockingAction("Stabs Re-import", getName()) {
            @Override
            public void actionPerformed(ActionContext context) {
                Program program = getCurrentProgram();
                if (program == null) {
                    Msg.showInfo(getClass(), null, "Stabs Re-import", "No program is open.");
                    return;
                }
                int tx = program.startTransaction("Stabs: clear done flag (re-import)");
                try {
                    program.getOptions(Program.PROGRAM_INFO)
                        .setBoolean(StabsAnalyzer.STABS_DONE_OPTION, false);
                } finally {
                    program.endTransaction(tx, true);
                }
                AutoAnalysisManager mgr = AutoAnalysisManager.getAnalysisManager(program);
                StabsAnalyzer analyzer = new StabsAnalyzer();
                mgr.scheduleOneTimeAnalysis(analyzer, program.getMemory());
            }

            @Override
            public boolean isEnabledForContext(ActionContext context) {
                Program p = getCurrentProgram();
                if (p == null) return false;
                return p.getMemory().getBlock(".stab") != null
                    && p.getMemory().getBlock(".stabstr") != null;
            }
        };
        reimport.setMenuBarData(new MenuData(new String[] { "&Tools", "Stabs", "&Re-import" }, null, "Stabs"));
        reimport.setEnabled(true);
        tool.addAction(reimport);
    }
}

File: src/main/kotlin/ghistabs/StabsPlugin.kt
package ghistabs

import docking.ActionContext
import docking.action.DockingAction
import docking.action.MenuData
import ghidra.app.CorePluginPackage
import ghidra.app.plugin.PluginCategoryNames
import ghidra.app.plugin.ProgramPlugin
import ghidra.app.plugin.core.analysis.AutoAnalysisManager
import ghidra.framework.plugintool.PluginInfo
import ghidra.framework.plugintool.PluginTool
import ghidra.framework.plugintool.util.PluginStatus
import ghidra.util.HelpLocation
import ghidra.util.Msg
import ghistabs.ImportOptions.Companion.markStabsDone
import ghistabs.parse.StabReader

/**
 * `Tools > Stabs > Re-import`: clears the persistent done-flag and re-runs the StabsAnalyzer.
 * The render is exported through [StabsDecompExporter] (`File > Export Program…`), not from here.
 */
@PluginInfo(
    status = PluginStatus.STABLE,
    packageName = CorePluginPackage.NAME,
    category = PluginCategoryNames.ANALYSIS,
    shortDescription = "Re-run the STABS importer on the current program.",
    description = "Adds 'Tools > Stabs > Re-import', which clears the 'Stabs Imported' flag and " +
        "re-runs auto-analysis so the StabsAnalyzer executes again.",
)
class StabsPlugin(tool: PluginTool) : ProgramPlugin(tool) {
    init {
        tool.addAction(
            object : DockingAction("Stabs Re-import", getName()) {
                override fun actionPerformed(context: ActionContext?) {
                    val program = currentProgram
                        ?: return Msg.showInfo(javaClass, null, "Stabs", "No program is open.")
                    program.markStabsDone(false)
                    AutoAnalysisManager.getAnalysisManager(program).reAnalyzeAll(null)
                }

                override fun isEnabledForContext(context: ActionContext?) =
                    currentProgram?.let(StabReader::hasStabs) == true
            }.apply {
                menuBarData = MenuData(arrayOf("&Tools", "Stabs", "&Re-import"), null, "Stabs")
                // `src/main/help/help/topics/Stabs/` — the topic dir name is the topic id.
                helpLocation = HelpLocation("Stabs", "Stabs_Reimport")
                isEnabled = true
            },
        )
    }
}
